ratis-server/src/main/java/org/apache/ratis/server/impl/TransferLeadership.java

```diff
@@ -81,11 +83,47 @@ public String toString() {
     this.server = server;
   }
 
+  private RaftPeerId getTransferee() {
+    return Optional.ofNullable(pending.get())
+        .map(r -> r.getRequest().getNewLeader()).orElse(null);
+  }
+
   boolean isSteppingDown() {
     return pending.get() != null;
   }
 
-  CompletableFuture<RaftClientReply> start(TransferLeadershipRequest request) {
+  void onFollowerAppendEntriesReply(LeaderStateImpl leaderState, FollowerInfo follower) {
+    final RaftPeerId transferee = getTransferee();
+    // If TransferLeadership is in progress, and the transferee has just append some entries
+    if (follower.getPeer().getId().equals(transferee)) {
```

Comment:
It is more efficient to check null first.  We may also move the `Optional` inside:
```java
  void onFollowerAppendEntriesReply(LeaderStateImpl leaderState, FollowerInfo follower) {
    final Optional<RaftPeerId> transferee = Optional.ofNullable(pending.get()).map(r -> r.getRequest().getNewLeader());

    // If TransferLeadership is in progress, and the transferee has just append some entries
    if (transferee.filter(t -> t.equals(follower.getId())).isPresent()) {
```